Better routing abstraction

tsareg opened the pull request:
Introduced ScreenRegistry - now app-level screens register themselves in this registry and App component knows nothing about specific screens.

Currently when we navigate to different part of application in React-Native we push a new scene into our navigator's scene stack: https://github.com/jitsi/jitsi-meet-react/blob/master/index.native.js#L27-39
Here is the article describing problem with this: https://blog.nativebase.io/the-caveats-of-using-navigator-in-react-native-9547d99172ce#.prwf1f48t

When we pushed views into the navigator, the navigator got stacked with new pages. As we kept pushing new views to the navigator, the app started to slow down and lag. 
The reason was that whenever a new view was pushed, it was mounted every time causing the app to slow down.
....
Navigating to new screen by push() doesn’t destroy the instance of the last screen and the component isn’t unmounted at all until it is popped out. So, if you’re not handling push() and pop() correctly, you may end-up in huge memory leak issues, making your app really slow.
It happened with us and we learnt how different Navigator is from other routers on the web.
So mainly each time we push new scene to navigator, it mounts a new scene, without cleaning the previous ones.
Instead in this PR it is suggested to replace current scene with new one (https://facebook.github.io/react-native/docs/navigator.html#replace).

Also in scope of this PR welcome page specific state that holds room name for now was introduced.

Some statistics: after join-quit conference for 4 times without this fix app consumes more than 100 MB of RAM. With fix - about 85 MB of RAM.

Introduced more "component-oriented design" - now components dispatch actions when they are mounted/unmounted instead of handling this inside navigation.

tsareg committed:
Create only two scenes that are not mounted/unmounted during navigation in mobile app.
Introduce welcome page specific state that holds room name for now.

tsareg committed:
Merge remote-tracking branch 'remotes/upstream/master' into mobile_navigation_no_additional_screens

tsareg committed:
Changes after merge

tsareg committed:
Introduced ScreenRegistry - now app-level screens register themselves in this registry and App component knows nothing about specific screens.
As much as possible unified 'onLeave' and 'onEnter' screen behavior for native and web routing.

legastero commented:
I think that the `ScreenRegistry` could be simplified by using a `Map` internally so that we can use the `for ... of` iterator in the places where `getAllScreens()` is used now.

Requiring the screens to include an `index` field to ensure consistent ordering by `getAllScreens()` doesn't seem right unless there's another use that I missed seeing.

tsareg commented:
@legastero 

> I think that the ScreenRegistry could be simplified by using a Map internally so that we can use the for ... of iterator in the places where getAllScreens() is used now.

I think we can store them not as Object or Map at all. I'll change it to Set internally.

> Requiring the screens to include an index field to ensure consistent ordering by getAllScreens() doesn't seem right unless there's another use that I missed seeing.

This is not for the ordering. It's used for React-Native's navigator - https://facebook.github.io/react-native/docs/navigator.html#additional-scenes. But I'll check if this is strictly required.

legastero commented:
> This is not for the ordering. It's used for React-Native's navigator - https://facebook.github.io/react-native/docs/navigator.html#additional-scenes. But I'll check if this is strictly required.

Aha! That bit of info clears a lot of this up for me :)

legastero commented on features/welcome/index.js:
```diff
@@ -1 +1,3 @@
+export * from './actions';
+export * from './constants';
 export * from './components';
```
Not quite sorted :)

tsareg replied:
It seems I need to learn English alphabet better :) For some reason I though N comes before M ))

legastero commented on features/toolbar/components/Toolbar.js:
```diff
@@ -1,10 +1,8 @@
-    navigate
-} from '../../app';
+import { navigate } from '../../base/navigation';
+import { WELCOME_SCREEN } from '../../welcome';
```
I know we don't want to import things from non-base features, but I'm not seeing a better option that doesn't lead to defining all of the screens inside the base directory.

legastero commented on features/conference/screen.js:
```diff
@@ -0,0 +1,52 @@
+const ConferenceScreen = {
+    component: Conference,
+    index: 1,
+    navigate: navigate,
```
Should we use the ES6 form here?

tsareg replied:
Will change, thanks

legastero commented on features/conference/index.js:
```diff
@@ -1,2 +1,4 @@
-export * from './components';
+import './screen';
 
+export * from './constants';
```
Sorting

tsareg replied:
It's import first, so I don't think we should put export before import.

tsareg committed:
Merge remote-tracking branch 'remotes/upstream/master' into routing_abstraction

tsareg committed:
Changes after code review:
- fixed import order
- removed 'index' property from Screens
- refactored ScreenRegistry to hold list of screens as Set

tsareg commented:
@legastero I've updated the PR according to your concerns

tsareg committed:
Merge remote-tracking branch 'remotes/upstream/master' into routing_abstraction

tsareg committed:
More Component-oriented design - now components dispatch actions when they are mounted/unmounted instead of handling this inside navigation.

tsareg committed:
Merge remote-tracking branch 'remotes/upstream/master' into routing_abstraction

tsareg committed:
Removed 'features/welcome' and moved roomName field in 'features/base/conference'

tsareg committed:
Do not dispatch ROOM_NAME_SET on every text change on welcome page, instead dispatch it when JOIN button is pressed

lyubomir commented:
@tsareg, please don't modify this PR from now on. Thank you!

lyubomir commented on features/conference/navigationHandler.native.js:
```diff
@@ -0,0 +1,19 @@
+    // TODO: currently replace method doesn't support animation, but work
+    // towards adding it is done in
+    // https://github.com/facebook/react-native/issues/1981
+    action.navigator.replace(route);
```
@tsareg, the change from the push method to the replace method is in its own (i.e. independently of the other modifications in this PR) the solution to the huge memory retention, right?

Initially, you started with the jumpTo method (as I remember you telling us in a meeting & I see in the first commit in this PR). Did you switch to the replace method later on in order to have exactly 1 mounted component/route/scene?

tsareg replied:
Yes. At first I thought of having both Conference and WelcomePage mounted all the time and jump between two of them. But later it turned out that it's not convenient. Instead in https://github.com/jitsi/jitsi-meet-react/pull/56/commits/75b1fbf8ae1c240311d291beed94b709249d35b7 I've introduced approach when each screen-level component initializes itself when is mounted and disposes itself when is unmounted. I'd say this approach is much closer to React/Redux principles then the previous one when we did something onEnter/onLeave.
